### PR TITLE
Basic structure for CMake + Azure Pipelines + QT

### PR DESCRIPTION
**Pull request by Necktschnagge**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,74 +1,83 @@
-cmake_minimum_required(VERSION 3.13.4)
```
Comment (Necktschnagge) [inline CMakeLists.txt:1]: Please decide for min version!

**Comment by Necktschnagge** [comment]
Almost done it, 
structure looks fine.

Still open:
 - [ ] add the library target for our sources.   _(Please define a global variable Project_name_lib,)_
 - [ ] Add the tests run by googletest.

File: CMakeLists.txt [hunk 2]
```diff
@@ -1,73 +1,85 @@
...
+set(CMAKE_AUTORCC ON)
 
-############### Language and Flags ###############
 set(CMAKE_CXX_STANDARD 17)
-set(CMAKE_CXX_STANDARD_REQUIRED True)
+set(CMAKE_CXX_STANDARD_REQUIRED ON)
 set(CMAKE_CXX_EXTENSIONS False)
-set(CMAKE_CXX_FLAGS_RELEASE "/MT /O2 /Ob2 /DNDEBUG")
-set(CMAKE_CXX_FLAGS_DEBUG "/MT /Zi /Ob0 /Od /RTC1")
+#set(CMAKE_CXX_FLAGS_RELEASE "/MT /O2 /Ob2 /DNDEBUG")
```
Comment (Necktschnagge) [inline CMakeLists.txt:48]: uncomment this

File: CMakeLists.txt (hunk 1 above)
Comment (Necktschnagge) [reply CMakeLists.txt:1]: @r0ost3r Please tell me what `cmake --version`s do you use on your local machines.
Is min_version 3.20 am option for you?

File: CMakeLists.txt [hunk 3]
```diff
@@ -3,6 +3,8 @@ cmake_minimum_required(VERSION 3.20)
 
 
 ############### Subdirectories for dependencies ###############
+set(CMAKE_CXX_FLAGS_RELEASE "/MT /O2 /Ob2 /DNDEBUG")
+set(CMAKE_CXX_FLAGS_DEBUG "/MT /Zi /Ob0 /Od /RTC1")
```
Comment (Necktschnagge) [inline CMakeLists.txt:7]: Please see these linkage errors:
https://dev.azure.com/Necktschnagge/341439fd-7b7d-482a-8ae7-6ae03646a61d/_apis/build/builds/11007/logs/60

The problem is about using dynamic vs static runtime library on windows. There are some flags to tell the compiler what to use.
CMake creates multiple targets, especially our Tobor_static_lib target as well as the binaries of googletest resulting from`add_subdirectory(extern/googletest)`.
The Test-Tobor target created in subdirectory test creates the executable Test-Tobor.exe.
This application must not use libraries which use different versions of multithreaded runtime library.
See Microsofts explainations here:
https://learn.microsoft.com/de-de/cpp/build/reference/md-mt-ld-use-run-time-library?view=msvc-170

May you figure out how to fix this @r0ost3r ?
If you gain any knowledge, please write it down, e.g. what is the default behaviour, which target needs to be modified?

**Comment by Necktschnagge** [comment]
Azure / ubuntu seems to have some update repository issuel, apt-get stops with some 404 error, unable to install Qt.
It is only on ubuntu 22, it is not affecting ubuntu 20.

It may be a good decision to just wait. It looks like some azure mirror is unavailable. Maybe Azure will fix this issue soon.

**Comment by Necktschnagge** [comment]
TODOs

- [ ] Please add some runtime test checking that QT5 / QT6 is used as expected.

File: CMakeLists.txt (hunk 3 above)
Comment (Necktschnagge) [reply CMakeLists.txt]: Currently Azure cannot load a bunch of dependencies.
Better wait some time and hope it will be working again soon.

Comment (Necktschnagge) [reply CMakeLists.txt]: Use this (since Cmake 3.15):
https://cmake.org/cmake/help/latest/prop_tgt/MSVC_RUNTIME_LIBRARY.html

Comment (Necktschnagge) [reply CMakeLists.txt]: I solved this issue.
using `cmake --verbose` the log shows us which targets used which flag (`/MTd` or `/MDd`):
**/MTd:** googletest(2 times googletest/googletest and 2 times googletest/googlemock)
**/MDd:** src/ (2 times, one for .exe, one for .lib) + test/

The solution is to [make googletest use MDd](https://stackoverflow.com/questions/12540970/how-to-make-gtest-build-mdd-instead-of-mtd-by-default-using-cmake)

**Comment by Necktschnagge** [comment]
fr review compare with https://stackoverflow.com/questions/14172856/compile-with-mt-instead-of-md-using-cmake

File: CMakeLists.txt [hunk 4]
```diff
@@ -1,73 +1,92 @@
...
-############### Language and Flags ###############
 set(CMAKE_CXX_STANDARD 17)
-set(CMAKE_CXX_STANDARD_REQUIRED True)
+set(CMAKE_CXX_STANDARD_REQUIRED ON)
 set(CMAKE_CXX_EXTENSIONS False)
-set(CMAKE_CXX_FLAGS_RELEASE "/MT /O2 /Ob2 /DNDEBUG")
-set(CMAKE_CXX_FLAGS_DEBUG "/MT /Zi /Ob0 /Od /RTC1")
+if (WIN32) #remove this lines?
+	set(CMAKE_CXX_FLAGS_RELEASE "/MT /O2 /Ob2 /DNDEBUG") #what do all these flags mean?
+	set(CMAKE_CXX_FLAGS_DEBUG "/MTd /Zi /Ob0 /Od /RTC1")
```
Comment (Necktschnagge) [inline CMakeLists.txt:53]: remove this lines if they are ignored

**Comment by Necktschnagge** [comment]
Todo
- [ ] Review
- [ ] Check for open TODOs of #1

File: CMakeLists.txt (hunk 4 above)
Comment (Necktschnagge) [reply CMakeLists.txt:53]: not to be resolved here. Created issue #19

File: CMakeLists.txt [hunk 5]
```diff
@@ -1,73 +1,92 @@
...
+
+############### Libraries via include_directories ###############
 include_directories(SYSTEM extern/spdlog/include)
-add_subdirectory(extern/googletest)
+
+
+
+############### Libraries via find_package ###############
+if(USE_QT6)
+	find_package(QT NAMES Qt6 Qt5 REQUIRED COMPONENTS Widgets Svg SvgWidgets)
```
Comment (Necktschnagge) [inline CMakeLists.txt:72]: Why do we list Qt5 if we operate on Qt6?

File: azure-pipelines.yml [hunk 6]
```diff
@@ -5,67 +5,151 @@ variables:
...
+         condition: and(succeeded(), eq( variables['Agent.OS'], 'Darwin'))
+       - script: |
+          echo "++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++"
+          pip install aqtinstall
+          echo "++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++"
+          aqt list-qt windows desktop
+          echo "++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++"
+          mkdir C:\Qt
+          echo "++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++"
+          aqt install-qt --outputdir C:\Qt windows desktop 5.15.2 win64_msvc2019_64
```
Comment (Necktschnagge) [inline azure-pipelines.yml:64]: Do windows 2019 / 2022 use different MSVC versions?

File: azure-pipelines.yml [hunk 7]
```diff
@@ -5,67 +5,151 @@ variables:
...
+      
+      pool:
+        vmImage: $(imageName)
+        name: $(the_name)
+      
+      steps:
+       - script: sudo apt-get update && sudo apt-get install build-essential libgl1-mesa-dev qt6-base-dev libqt6svg6 libqt6svg6-dev
+         displayName: "Install Qt (Linux)"
+         condition: and(succeeded(), eq( variables['Agent.OS'], 'Linux'))
+       - script: brew install qt@5 qt@6
```
Comment (Necktschnagge) [inline azure-pipelines.yml:120]: why qt@5 ?

File: azure-pipelines.yml (hunk 6 above)
Comment (Necktschnagge) [reply azure-pipelines.yml:64]: What are clang / gcc versions on Linux / Mac ?

File: src/main.cpp [hunk 8]
```diff
@@ -1,8 +1,16 @@
...
+#include "mainwindow.h"
+
+#include <QApplication>
+
+
+int main(int argc, char *argv[])
 {
 	init_logger();
-	return 0;
+    QApplication a(argc, argv);
```
Comment (Necktschnagge) [inline src/main.cpp:12]: What are the side effects of passing comandline parameters?

File: src/CMakeLists.txt [hunk 9]
```diff
@@ -1,13 +1,122 @@
...
 		pthread
 	)
 endif()
 
-add_library(${PROJECT_NAME}_lib STATIC ${SOURCES}) #for using in tests
+
+
+############### Target Properties ###############
+set_target_properties(${PROJECT_NAME} PROPERTIES
+	MACOSX_BUNDLE_GUI_IDENTIFIER my.example.com
```
Comment (Necktschnagge) [inline src/CMakeLists.txt:48]: What are all these properties for?

File: src/mainwindow.cpp [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+</svg>
+)xxx"
+    };
+
+    QXmlStreamReader xml;
+    xml.addData(example_svg_string);
+
+    QSvgRenderer* svgRenderer = new QSvgRenderer(&xml);
+    QGraphicsSvgItem *item = new QGraphicsSvgItem();
+    QGraphicsScene *scene = new QGraphicsScene();
```
Comment (Necktschnagge) [inline src/mainwindow.cpp:39]: This looks like two lines of creating memory leaks.
Check ownership of these pointers.

File: test/CMakeLists.txt [hunk 11]
```diff
@@ -1,18 +1,21 @@
...
+
+############### Target via add_executable ###############
 add_executable(${TEST_PROJECT_NAME} ${TEST_SOURCES})
 
-add_test(NAME ${TEST_PROJECT_NAME} COMMAND ${TEST_PROJECT_NAME})
 
-############### Libraries ###############
+
+############### Test via add_test ###############
+add_test(NAME ${TEST_PROJECT_NAME} COMMAND ${TEST_PROJECT_NAME}) # what does command mean in detail?
```
Comment (Necktschnagge) [inline test/CMakeLists.txt:11]: check what COMMAND is about

**Review by Necktschnagge — COMMENTED** [review]
self-review done 🆗 

Almost all comments I left may be treated as action items for later.

File: CMakeLists.txt (hunk 5 above)
Comment (Necktschnagge) [reply CMakeLists.txt:72]: left open for #20

**Comment by Necktschnagge** [comment]
@r0ost3r Please review if you have time and write down all inconsistencies, mistakes, etc. you discover.

I've completed my review.
Most comments I made are pure cosmetical issues on such a level so that I will not investigate them here.
It can be done later.

However, please note everything that looks not okay or confusing.
We'll fix inmortant thing and keep others in mind for later.

File: azure-pipelines.yml (hunk 7 above)
Comment (Necktschnagge) [reply azure-pipelines.yml]: Removed qt5 there

File: src/main.cpp (hunk 8 above)
Comment (Necktschnagge) [reply src/main.cpp:12]: to be continued in #22

File: src/mainwindow.cpp (hunk 10 above)
Comment (Necktschnagge) [reply src/mainwindow.cpp:39]: see #23

File: src/CMakeLists.txt (hunk 9 above)
Comment (Necktschnagge) [reply src/CMakeLists.txt:48]: see #24

File: test/CMakeLists.txt (hunk 11 above)
Comment (Necktschnagge) [reply test/CMakeLists.txt:11]: see #25

File: CMakeLists.txt (hunk 1 above)
Comment (r0ost3r) [reply CMakeLists.txt:1]: cmake version 3.16.3

**Review by r0ost3r — CHANGES_REQUESTED** [review]
src/main.cpp
src/mainwindow.cpp
src/mainwindow.h
src/mainwindow.ui
sind doof.

**Comment by Necktschnagge** [comment]
@r0ost3r we replace ui code in next Pull.

**Review by r0ost3r — APPROVED** [review]

**Comment by Necktschnagge** [comment]
@r0ost3r I added the setup steps for README.md like you already saw yesterday.

Please let me know your final conclusion after checking for unnecessary packages.
Please let your comments here or approve review.

**Comment by Necktschnagge** [comment]
Todo
- [x] Review
- [x] Check for open TODOs of #1 
- [ ] Check if running on local development platform works for
   - [x] Windows Qt5
   - [x] Windows Qt6
   - [ ] Windows Qt Creator (5 + 6)
   - [x] Linux Qt5 (skipped, like azure)
   - [x] Linux Qt6 (skipped, like azure)
   - [ ] Linux Qt-Creator ?
- [x] Add readme steps of setting up development environment. (windows Path variable, QT_DIR, ...)
- runtime check like in #1 should be done via ci jobs. 🕚 **postponed**
- [ ] check that one can run Tobor.exe + Test-Tobor.exe from Visual Studio in both Debug and Release mode.

**Review by r0ost3r — DISMISSED** [review]

**Review by r0ost3r — APPROVED** [review]

**Pull request merged**